regression/contracts/loop_assigns-03/test.desc

```diff
@@ -0,0 +1,17 @@
+CORE
+main.c
+--apply-loop-contracts
+^EXIT=10$
+^SIGNAL=0$
+^\[main.\d+\] .* Check loop invariant before entry: SUCCESS$
+^\[main.\d+\] .* Check that i is assignable: FAILURE$
+^\[main.\d+\] .* Check that b->data\[(.*)i\] is assignable: SUCCESS$
+^VERIFICATION FAILED$
+--
+--
+This test (taken from #6021) shows the need for assigns clauses on loops.
+The alias analysis in this case returns `unknown`,
+and so we must manually annotate an assigns clause on the loop.
+
+Note that the annotated assigns clause in this case is an underapproximation,
+because `i` is also assigned in the loop and should be marked as assignable.
```

Comment:
but again, isn't that part of the "local write set"?

Reply:
There's no distinction between "local" and "global". It's all writable memory visible at the current scope.

Reply:
moreover the information that `i` was declared `for( <here>; ... ; ...)` is destroyed by C-to-goto conversion

Reply:
> moreover the information that `i` was declared `for( <here>; ... ; ...)` is destroyed by C-to-goto conversion

Oh that clarifies this a bit more. So a simple loop like this
```c
  for(int i = 0; i < 2; i++)
    assert(i < 2);
```
is converted to something like this
```
        // 0 file main.c line 13 function main
        DECL main::1::1::i : signedbv[32]
        // 1 file main.c line 13 function main
        ASSIGN main::1::1::i := 0
        // 2 file main.c line 13 function main
     1: IF ¬(main::1::1::i < 2) THEN GOTO 2
        // 3 file main.c line 14 function main
        ASSERT main::1::1::i < 2 // assertion i < 2
        // 4 file main.c line 13 function main
        ASSIGN main::1::1::i := main::1::1::i + 1
        // 5 file main.c line 13 function main
        GOTO 1
        // 6 file main.c line 14 function main
     2: DEAD main::1::1::i
```

So the declaration of `i` in this context is outside of the loop body. If we have local variables in the loop body, do users must add them to the assign clause as well? It's a bit odd that we don't have access to these initial declarations since their scope is only the loop body.

Reply:
> If we have local variables in the loop body, do users must add them to the assign clause as well?

Local variables (variables declared inside the loop / function body) are completely unrelated here. Loop-local variables need not be specified in the assigns clause -- they aren't even visible at the enclosing scope.

I think the main confusion here is that `i` is "a local variable in the loop body". It's not; it's outside of the loop body.

Before C99 standard, declarations within `for` loop initialization were not allowed. So the declaration for `i` had to be necessarily made outside of the loop. Starting with C99, `i` can be declared in the initialization part of the `for` loop and has the same scope as the loop (not the loop body).

[cppreference states that](https://en.cppreference.com/w/cpp/language/for):
```c
for ( init-statement ; condition ; iteration-expression ) statement
```
is equivalent to: (comments added by me) 
```c
{
    // This is NOT the loop-body scope
    init-statement
    while ( condition )
    // This is where the invariant goes
    {
        // This is the loop-body scope
        statement
        iteration-expression ;
    } 
} 
```

Reply:
Thank you so much for adding a bit more context here! If there is a "loop scope" that encapsulates the entire loop and loop body, why are we focusing only in the loop body and not the most general scope? The only goal I'm trying to active with this discussion is: decrease the burden of annotation for users (this is not a blocker for this PR).

Reply:
Your previous comment respond my previous question: https://github.com/diffblue/cbmc/pull/6249#discussion_r749646842. If you could grab all declarations from the initialization clause of a `for` loop and add them to the assigns clause automatically, this would decrease the annotation burden to developers and it's perfectly sound for our approach. Thanks!

Reply:
Yes exactly. I'll try to push that change in this PR if it works (still playing with it), or would make a separate PR for it later.